### PR TITLE
[IOTDB-850] check list user/role privilege 

### PR DESCRIPTION
**Pull request by haimeiguo**
user can only LIST USER or LIST ROLE if he is granted list_user / list_role privilege, check if he is granted in plan executor

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/qp/physical/PhysicalPlan.java [hunk 1]
```diff
@@ -48,6 +48,7 @@
 
   private static final String SERIALIZATION_UNIMPLEMENTED = "serialization unimplemented";
 
+  protected String userName;
```
Comment (neuyilan) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/PhysicalPlan.java:51]: Does every physical plan have to have a userName?
for OOP, I do not think its a good design.

**Review by neuyilan — CHANGES_REQUESTED** [review]
Title:
850 - check list user privilege -> [IOTDB-850] check list user privilege

File: server/src/main/java/org/apache/iotdb/db/qp/physical/PhysicalPlan.java (hunk 1 above)
Comment (haimeiguo) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/PhysicalPlan.java:51]: Yes, I will change it to AuthorPlan only. Thank you for your comment!

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 2]
```diff
@@ -1059,7 +1060,7 @@ protected boolean deleteStorageGroups(DeleteStorageGroupPlan deleteStorageGroupP
     }
     return true;
   }
-
+
```
Comment (mychaow) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1063]: delete the blank space.

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAuthorizationIT.java [hunk 3]
```diff
@@ -793,15 +793,19 @@ public void testListRolePrivileges() throws ClassNotFoundException, SQLException
...
+      ResultSet resultSet = adminStmt.executeQuery("LIST ROLE PRIVILEGES role1");
+      String ans = "";
+      try {
+        //not granted list role privilege, should return empty
+        validateResultSet(resultSet, ans);
+
+        adminStmt.execute(
           "GRANT ROLE role1 PRIVILEGES 'READ_TIMESERIES','INSERT_TIMESERIES','DELETE_TIMESERIES' ON root.a.b.c");
-      adminStmt.execute(
+        adminStmt.execute(
```
Comment (mychaow) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAuthorizationIT.java:804]: delete the blank space.

**Review by mychaow — CHANGES_REQUESTED** [review]
Good job!

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 4]
```diff
@@ -1116,14 +1125,26 @@ private ListDataSet executeListRole() {
...
     typeList.add(TSDataType.TEXT);
-    int index = 0;
     ListDataSet dataSet = new ListDataSet(headerList, typeList);
+
+    if (!hasListUserPrivilege) {
+      return dataSet;
+    }
+
+    List<String> userList = authorizer.listAllUsers();
+    int index = 0;
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1146]: What is this index for?

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 5]
```diff
@@ -1116,14 +1125,26 @@ private ListDataSet executeListRole() {
     return dataSet;
   }
 
-  private ListDataSet executeListUser() {
-    List<String> userList = authorizer.listAllUsers();
+  private ListDataSet executeListUser(AuthorPlan plan) throws AuthException {
+    List<PartialPath> paths = new ArrayList<>();
+    paths.add(plan.getNodeName());
+    // check if current user is granted list_user privilege
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1131]: Better replaced with `Collections.singletonList()`

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAuthorizationIT.java (hunk 3 above)
Comment (haimeiguo) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBAuthorizationIT.java]: done. Thank you :)

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 6]
```diff
@@ -1098,13 +1099,21 @@ protected QueryDataSet processAuthorQuery(AuthorPlan plan)
...
+  private ListDataSet executeListRole(AuthorPlan plan) throws AuthException {
     int index = 0;
     List<PartialPath> headerList = new ArrayList<>();
     List<TSDataType> typeList = new ArrayList<>();
     headerList.add(new PartialPath(COLUMN_ROLE, false));
     typeList.add(TSDataType.TEXT);
     ListDataSet dataSet = new ListDataSet(headerList, typeList);
+
+    boolean hasListRolePrivilege = AuthorityChecker
+        .check(plan.getLoginUserName(), new ArrayList<>(), plan.getOperatorType(), plan.getLoginUserName());
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1111]: Better replaced with `Collections.emptyList()`.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/sys/AuthorPlan.java [hunk 7]
```diff
@@ -39,12 +39,13 @@
...
 
   private AuthorOperator.AuthorType authorType;
-  private String userName;
   private String roleName;
   private String password;
   private String newPassword;
   private Set<Integer> permissions;
   private PartialPath nodeName;
+  private String userName;
+  private String loginUserName;
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/sys/AuthorPlan.java:48]: I think maybe we can push loginUserName down to PhysicalPlan. It is often meaningful to know who issued this plan.

**Review by jt2594838 — CHANGES_REQUESTED** [review]
Good, but pay attention to some coding tricks.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/sys/AuthorPlan.java (hunk 7 above)
Comment (haimeiguo) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/sys/AuthorPlan.java]: done. thank you! 
index is for rowRocord, it was there before my edit.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/PhysicalPlan.java (hunk 1 above)
Comment (jt2594838) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/PhysicalPlan.java]: I have a different opinion. Every physical plan is issued by a user and keeping this information will help us tracking user behaviors in the future.

**Review by jt2594838 — APPROVED** [review]

**Review by mychaow — APPROVED** [review]

**Pull request merged**